File: Source/Aladdin.CAPI.GOST/CAPI/GOST/Mode/GOSTR3412/CBC.cs
using System; 

namespace Aladdin.CAPI.GOST.Mode.GOSTR3412
{
    ///////////////////////////////////////////////////////////////////////////////
    // Режим CBC
    ///////////////////////////////////////////////////////////////////////////////
    public class CBC : CAPI.Mode.CBC
    {
        // режим смены ключа и размер смены ключа
        private KeyDerive keyMeshing; private int N; 
    
        // конструктор
	    public CBC(CAPI.Cipher engine, CipherMode.CBC parameters, 
            KeyDerive keyMeshing, int N, PaddingMode padding) : base(engine, parameters, padding)
	    { 
            // проверить корректность параметров
            if ((N % engine.BlockSize) != 0) throw new ArgumentException(); 

            // сохранить переданные параметры
            this.keyMeshing = RefObject.AddRef(keyMeshing); this.N = N; 
	    }
        // конструктор
	    public CBC(CAPI.Cipher engine, CipherMode.CBC parameters, KeyDerive keyMeshing, int N)
            
            // сохранить переданные параметры
            : this(engine, parameters, keyMeshing, N, PaddingMode.Any) {}

        // конструктор
	    public CBC(CAPI.Cipher engine, CipherMode.CBC parameters, PaddingMode padding) 
            
            // сохранить переданные параметры
            : base(engine, parameters, padding) { this.keyMeshing = null; N = 0; }

        // конструктор
	    public CBC(CAPI.Cipher engine, CipherMode.CBC parameters) 
            
            // сохранить переданные параметры
            : this(engine, parameters, PaddingMode.Any) {}

        // деструктор
        protected override void OnDispose() 
        { 
            // освободить ресурсы
            RefObject.Release(keyMeshing); base.OnDispose();
        }
        // преобразование зашифрования
        protected override Transform CreateEncryption(ISecretKey key) 
        { 
            // преобразовать тип параметров
            CipherMode.CBC parameters = (CipherMode.CBC)Mode; 

            // преобразование зашифрования
            return new CBC_ENC(Engine, keyMeshing, N, key, parameters); 
        }
        // преобразование расшифрования
        protected override Transform CreateDecryption(ISecretKey key) 
        { 
            // преобразовать тип параметров
            CipherMode.CBC parameters = (CipherMode.CBC)Mode; 

            // преобразование расшифрования
            return new CBC_DEC(Engine, keyMeshing, N, key, parameters); 
        }
    }
}

File: Source/Aladdin.CAPI.GOST.Java/src/aladdin/capi/gost/mode/gost28147/CBC.java
package aladdin.capi.gost.mode.gost28147;
import aladdin.*; 
import aladdin.capi.*; 
import java.security.*; 
import java.io.*; 

///////////////////////////////////////////////////////////////////////////////
// Режим CBC
///////////////////////////////////////////////////////////////////////////////
public class CBC extends aladdin.capi.mode.CBC
{
    // режим смены ключа
    private final KeyDerive keyMeshing;
    
    // конструктор
	public CBC(Cipher engine, CipherMode.CBC parameters, KeyDerive keyMeshing, PaddingMode padding)
	{ 
        // сохранить переданные параметры
        super(engine, parameters, padding); this.keyMeshing = RefObject.addRef(keyMeshing); 
	}
	public CBC(Cipher engine, CipherMode.CBC parameters, KeyDerive keyMeshing)
	{ 
        // сохранить переданные параметры
        this(engine, parameters, keyMeshing, PaddingMode.ANY); 
	}
    // конструктор
	public CBC(Cipher engine, CipherMode.CBC parameters, PaddingMode padding)
	{ 
        // сохранить переданные параметры
        super(engine, parameters, padding); this.keyMeshing = null; 
	}
    // конструктор
    // конструктор
	public CBC(Cipher engine, CipherMode.CBC parameters)
	{ 
        // сохранить переданные параметры
        this(engine, parameters, PaddingMode.ANY); 
	}
    // деструктор
    @Override protected void onClose() throws IOException    
    { 
        // освободить ресурсы
        RefObject.release(keyMeshing); super.onClose();
    }
    // преобразование зашифрования
    @Override protected Transform createEncryption(ISecretKey key) 
    { 
        // преобразование зашифрования
        return new Encryption(engine(), keyMeshing, key, mode()); 
    }
    // преобразование расшифрования
    @Override protected Transform createDecryption(ISecretKey key) 
    { 
        // преобразование расшифрования
        return new Decryption(engine(), keyMeshing, key, mode()); 
    }
    ///////////////////////////////////////////////////////////////////////////////
    // Режим зашифрования CBC
    ///////////////////////////////////////////////////////////////////////////////
    public static class Encryption extends aladdin.capi.mode.CBC.Encryption
    {
        // алгоритм смены ключа и размер смены ключа
        private final KeyDerive keyMeshing; private final int N;
        // текущий ключ и накопленный размер
        private ISecretKey currentKey; private int length; 

        // конструктор
        public Encryption(Cipher engine, KeyDerive keyMeshing, ISecretKey key, CipherMode.CBC parameters) 
        { 
            // сохранить переданные параметры
            super(engine, key, parameters); currentKey = RefObject.addRef(key); 
            
            // сохранить переданные параметры
            this.keyMeshing = RefObject.addRef(keyMeshing); 
            
            // указать размер смены ключа
            N = (keyMeshing != null) ? 1024 : 0; 
        } 
        // конструктор
        public Encryption(Cipher engine, ISecretKey key, CipherMode.CBC parameters) 
        { 
            // смена ключа отсутствует
            super(engine, key, parameters); keyMeshing = null; N = 0; 
            
            // сохранить переданные параметры
            currentKey = RefObject.addRef(key); 
        } 
        // освободить ресурсы
        @Override protected void onClose() throws IOException    
        {
            // освободить ресурсы
            RefObject.release(currentKey);
            
            // освободить ресурсы
            RefObject.release(keyMeshing); super.onClose();
        }
        // инициализировать алгоритм
        @Override public void init() throws IOException { super.init(); length = 0; }
        
        @Override protected void update( 
            byte[] data, int dataOff, byte[] buf, int bufOff) throws IOException
        {
            // обработать полный блок
            super.update(data, dataOff, buf, bufOff); 
            
            // увеличить размер данных
            length += blockSize(); if (N == 0 || (length % N) != 0) return; 

            // изменить значение ключа
            try (ISecretKey key = keyMeshing.deriveKey(currentKey, iv(), currentKey.keyFactory(), 32))
            {
                // переустановить ключ
                if (key != currentKey) resetKey(key); 

                // сохранить новый текущий ключ
                RefObject.release(currentKey); currentKey = RefObject.addRef(key); 
            }
            // обработать неожидаемое исключение
            catch (InvalidKeyException e) { throw new RuntimeException(e); }
        }
    }
    ///////////////////////////////////////////////////////////////////////////////
    // Режим расшифрования CBC
    ///////////////////////////////////////////////////////////////////////////////
    public static class Decryption extends aladdin.capi.mode.CBC.Decryption
    {
        // алгоритм смены ключа и размер смены ключа
        private final KeyDerive keyMeshing; private final int N;
        // текущий ключ и накопленный размер
        private ISecretKey currentKey; private int length; 

        // конструктор
        public Decryption(Cipher engine, KeyDerive keyMeshing, ISecretKey key, CipherMode.CBC parameters) 
        { 
            // сохранить переданные параметры
            super(engine, key, parameters); currentKey = RefObject.addRef(key); 
            
            // сохранить переданные параметры
            this.keyMeshing = RefObject.addRef(keyMeshing); 

            // указать размер смены ключа
            N = (keyMeshing != null) ? 1024 : 0; 
        } 
        // конструктор
        public Decryption(Cipher engine, ISecretKey key, CipherMode.CBC parameters) 
        { 
            // смена ключа отсутствует
            super(engine, key, parameters); keyMeshing = null; N = 0; 
            
            // сохранить переданные параметры
            currentKey = RefObject.addRef(key); 
        } 
        // освободить ресурсы
        @Override protected void onClose() throws IOException    
        {
            // освободить ресурсы
            RefObject.release(currentKey);
            
            // освободить ресурсы
            RefObject.release(keyMeshing); super.onClose();
        }
        // инициализировать алгоритм
        @Override public void init() throws IOException { super.init(); length = 0; }
        
        @Override protected void update( 
            byte[] data, int dataOff, byte[] buf, int bufOff) throws IOException
        {
            // обработать полный блок
            super.update(data, dataOff, buf, bufOff); 
            
            // увеличить размер данных
            length += blockSize(); if (N == 0 || (length % N) != 0) return; 

            // изменить значение ключа
            try (ISecretKey key = keyMeshing.deriveKey(currentKey, iv(), currentKey.keyFactory(), 32))
            {
                // переустановить ключ
                if (key != currentKey) resetKey(key); 

                // сохранить новый текущий ключ
                RefObject.release(currentKey); currentKey = RefObject.addRef(key); 
            }
            // обработать неожидаемое исключение
            catch (InvalidKeyException e) { throw new RuntimeException(e); }
        }
    }
}